### PR TITLE
[GabTV] Improve extractor

### LINKED ISSUE #3416: GabTV channel_name error download
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

Brazil

### Description

When I try to use yt-dlp on GabTV an error is given: unable to extract channel_name 

### Verbose log

```shell
yt-dlp -vU https://tv.gab.com/channel/davecullen/view/three-experiments-that-explain-conformity-and-6254454f4c1b91f9610e8aa1
[debug] Command-line config: ['-vU', 'https://tv.gab.com/channel/davecullen/view/three-experiments-that-explain-conformity-and-6254454f4c1b91f9610e8aa1'][debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, err UTF-8, pref UTF-8
[debug] yt-dlp version 2022.04.08 [7884ade65] (zip)[debug] Python version 3.6.9 (CPython 64bit) - Linux-4.15.0-151-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 3.4.8, ffprobe 3.4.8, phantomjs ., rtmpdump 2.4
[debug] Optional libraries: brotli, certifi, Crypto, secretstorage, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.04.08, Current version: 2022.04.08
yt-dlp is up to date (2022.04.08)
[debug] [GabTV] Extracting URL: https://tv.gab.com/channel/davecullen/view/three-experiments-that-explain-conformity-and-6254454f4c1b91f9610e8aa1
[GabTV] 6254454f4c1b91f9610e8aa1: Downloading webpage
ERROR: [GabTV] three-experiments-that-explain-conformity-and-6254454f4c1b91f9610e8aa1: Unable to extract channel_name; please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
  File "./yt-dlp/yt_dlp/extractor/common.py", line 641, in extract
    ie_result = self._real_extract(url)
  File "./yt-dlp/yt_dlp/extractor/gab.py", line 36, in _real_extract
    channel_name = self._search_regex(r'data-channel-name=\"(?P<channel_id>[^\"]+)', webpage, 'channel_name')
  File "./yt-dlp/yt_dlp/extractor/common.py", line 1229, in _search_regex
    raise RegexNotFoundError('Unable to extract %s' % _name
```

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Close #3416

### CODE REVIEW

File: yt_dlp/extractor/gab.py [hunk 1]
```diff
@@ -1,13 +1,14 @@
 import re
 
 from .common import InfoExtractor
+from ..compat import compat_urlparse
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:4]:
```suggestion
import urllib.parse
```

File: yt_dlp/extractor/gab.py [hunk 2]
```diff
@@ -22,48 +23,52 @@ class GabTVIE(InfoExtractor):
...
+        sources = re.findall(r'<source.*?src=\"(?P<src>[^\"]+).+?size=\"(?P<quality>\d+)\"', webpage)
+        for source in sources:
+            if not source:
+                continue
+            formats.append({
+                'format_id': source[1],
+                'quality': source[1],
+                'ext': 'mp4',
+                'url': compat_urlparse.urljoin(self._DOMAIN, source[0]),
+            })
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:61]:
```suggestion
        formats = [{
            'format_id': q,
            'quality': q,
            'ext': 'mp4',
            'url': compat_urlparse.urljoin(self._DOMAIN, src),
        } for src, q in re.findall(r'<source[^>]+src="(?P<src>[^"]+)"[^>]*\ssize="(?P<quality>\d+)"', webpage)]
```

File: yt_dlp/extractor/gab.py [hunk 3]
```diff
@@ -22,48 +23,52 @@ class GabTVIE(InfoExtractor):
...
         return {
             'id': id,
             'title': title,
             'formats': formats,
             'description': description,
-            'uploader': channel_name,
+            'uploader': video_info.get('artist'),
             'uploader_id': channel_id,
-            'thumbnail': f'https://tv.gab.com/image/{id}',
+            'thumbnail': compat_urlparse.urljoin(self._DOMAIN, self._search_regex(r'data-poster=\"(?P<data_poster>[^\"]+)', webpage, 'data_poster')),
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:71]: should be non-fatal

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/gab.py [hunk 4]
```diff
@@ -22,48 +23,47 @@ class GabTVIE(InfoExtractor):
...
+    _DOMAIN = 'https://tv.gab.com/'
 
     def _real_extract(self, url):
         id = self._match_id(url).split('-')[-1]
         webpage = self._download_webpage(url, id)
         channel_id = self._search_regex(r'data-channel-id=\"(?P<channel_id>[^\"]+)', webpage, 'channel_id')
-        channel_name = self._search_regex(r'data-channel-name=\"(?P<channel_id>[^\"]+)', webpage, 'channel_name')
-        title = self._search_regex(r'data-episode-title=\"(?P<channel_id>[^\"]+)', webpage, 'title')
-        view_key = self._search_regex(r'data-view-key=\"(?P<channel_id>[^\"]+)', webpage, 'view_key')
+        video_info = self._parse_json(self._search_regex(r'(?s)mediaMetadata\:\s+(\{.+?\})', webpage, 'video_info'),
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:46]: Unnecessary escaping

File: yt_dlp/extractor/gab.py [hunk 5]
```diff
@@ -22,48 +23,47 @@ class GabTVIE(InfoExtractor):
...
+                                      video_id=id, transform_source=js_to_json)
+        title = video_info.get('title')
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:48]: should be directly in `return`

File: yt_dlp/extractor/gab.py [hunk 6]
```diff
@@ -22,48 +23,47 @@ class GabTVIE(InfoExtractor):
...
         return {
             'id': id,
             'title': title,
             'formats': formats,
             'description': description,
-            'uploader': channel_name,
+            'uploader': video_info.get('artist'),
             'uploader_id': channel_id,
-            'thumbnail': f'https://tv.gab.com/image/{id}',
+            'thumbnail': urllib.parse.urljoin(self._DOMAIN, self._search_regex(r'data-poster=\"(?P<data_poster>[^\"]+)', webpage, 'data_poster', fatal=False)),
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:66]: this is still effectively fatal. If the regex doesnt match, it becomes `urllib.parse.urljoin(self._DOMAIN, None)`

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/gab.py (hunk 4 above)
Comment (hatienl0i261299) [reply yt_dlp/extractor/gab.py:46]: @pukkandan You mean we dont need transform_source, is it right?

Comment (pukkandan) [reply yt_dlp/extractor/gab.py:46]: No, I don't know if transform_source is needed or not. You'll need to check that. 

I meant this:
```diff
-        video_info = self._parse_json(self._search_regex(r'(?s)mediaMetadata\:\s+(\{.+?\})', webpage, 'video_info'),
+        video_info = self._parse_json(self._search_regex(r'(?s)mediaMetadata:\s+({.+?})', webpage, 'video_info'),

File: yt_dlp/extractor/gab.py (hunk 6 above)
Comment (pukkandan) [reply yt_dlp/extractor/gab.py:66]:
```py
poster = self._search_regex(r'data-poster="(?P<data_poster>[^"]+)', webpage, 'poster', fatal=False)
...
'thumbnail': urllib.parse.urljoin(self._DOMAIN, poster) if poster else None
```

File: yt_dlp/extractor/gab.py [hunk 7]
```diff
@@ -22,48 +23,46 @@ class GabTVIE(InfoExtractor):
...
-            }
-            if 'audio-' in resolution:
-                frmt['abr'] = str_to_int(resolution.replace('audio-', ''))
-                frmt['height'] = 144
-                frmt['quality'] = -10
-            else:
-                frmt['height'] = str_to_int(resolution.replace('p', ''))
-            formats.append(frmt)
+        video_info = self._parse_json(self._search_regex(r'(?s)mediaMetadata:\s+({.+?})', webpage, 'video_info'),
+                                      video_id=id, transform_source=js_to_json)
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:47]: Use _search_json

File: yt_dlp/extractor/gab.py [hunk 8]
```diff
@@ -22,48 +23,46 @@ class GabTVIE(InfoExtractor):
...
+        formats = [{
+            'format_id': q,
+            'quality': q,
+            'ext': 'mp4',
+            'url': urllib.parse.urljoin(self._DOMAIN, src),
+        } for src, q in re.findall(r'<source[^>]+src="(?P<src>[^"]+)"[^>]*\ssize="(?P<quality>\d+)"', webpage)]
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:53]: What is `q`? width? height?

File: yt_dlp/extractor/gab.py [hunk 9]
```diff
@@ -22,48 +23,46 @@ class GabTVIE(InfoExtractor):
...
+        poster = self._search_regex(r'data-poster="(?P<data_poster>[^"]+)', webpage, 'poster', fatal=False)
+
         return {
             'id': id,
-            'title': title,
+            'title': video_info.get('title'),
             'formats': formats,
-            'description': description,
-            'uploader': channel_name,
+            'description': clean_html(self._html_search_regex(self._meta_regex('description'), webpage, 'description', group='content')) or None,
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:62]:
```suggestion
            'description': self._html_search_meta('description'),
```

File: yt_dlp/extractor/gab.py [hunk 10]
```diff
@@ -22,48 +23,46 @@ class GabTVIE(InfoExtractor):
...
+            'uploader': video_info.get('artist'),
             'uploader_id': channel_id,
-            'thumbnail': f'https://tv.gab.com/image/{id}',
+            'thumbnail': urllib.parse.urljoin(self._DOMAIN, poster) if poster else None,
```
Comment (pukkandan) [inline yt_dlp/extractor/gab.py:65]:
```suggestion
            'thumbnail': urljoin('https://tv.gab.com/', poster),
```
Use the urljoin from utils

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request closed without merging**